### PR TITLE
SbomExtractor improvements

### LINKED ISSUE #51: include component hashes
v0.3.0 does not add the hashes for the components (ie the dependent jars).

Examples can be seen in https://github.com/CycloneDX/bom-examples/blob/master/SBOM/dropwizard-1.3.15/bom.xml -- I'm not sure that all the hashes in the example are needed.

Some hashes should already be in the Coursier cache for each jar - Maven Central typically adds MD5 and SHA1. SBT or Coursier may provide support for getting these hashes or generating more.

### PR DESCRIPTION
**Pull request by lhns**
This PR adds the following attributes to sboms:
- plugin information (name, version)
- bom ref
- hashes (fixes #51)
- license id

It also fixes an error caused by modules that are reported multiple times (cross builds, complex dependsOn structure and such)

### CODE REVIEW

File: src/main/scala/com/github/sbt/sbom/BomExtractor.scala [hunk 1]
```diff
@@ -18,10 +19,27 @@ class BomExtractor(settings: BomExtractorParams, report: UpdateReport, log: Logg
...
       bom.setSerialNumber(serialNumber)
     }
+    if (settings.schemaVersion.getVersion >= Version.VERSION_12.getVersion) {
+      bom.setMetadata(metadata)
+    }
     bom.setComponents(components.asJava)
     bom
   }
 
+  private def metadata: Metadata = {
```
Comment (raboof) [inline src/main/scala/com/github/sbt/sbom/BomExtractor.scala:29]: maybe even `lazy val` or simply `val`?

File: src/main/scala/com/github/sbt/sbom/BomExtractor.scala [hunk 2]
```diff
@@ -18,10 +19,27 @@ class BomExtractor(settings: BomExtractorParams, report: UpdateReport, log: Logg
...
+    val metadata = new Metadata()
+    metadata.addTool(tool)
+    metadata
+  }
+
+  private def tool: Tool = {
+    val tool = new Tool()
+    // https://github.com/devops-kung-fu/bomber searches for string CycloneDX to detect format
```
Comment (raboof) [inline src/main/scala/com/github/sbt/sbom/BomExtractor.scala:37]: weird so good to know :). a more detailed reference would be even nicer if you have one.

File: src/main/scala/com/github/sbt/sbom/BomExtractor.scala [hunk 3]
```diff
@@ -77,13 +98,16 @@ class BomExtractor(settings: BomExtractorParams, report: UpdateReport, log: Logg
       component.setPurl(
         new PackageURL(PackageURL.StandardTypes.MAVEN, group, name, version, new util.TreeMap(), null).canonicalize()
       )
+      if (settings.schemaVersion.getVersion >= Version.VERSION_11.getVersion) {
+        component.setBomRef(component.getPurl)
+      }
       component.setScope(Component.Scope.REQUIRED)
-      licenseChoice.foreach(component.setLicenseChoice)
+      component.setHashes(hashes(artifactPaths(moduleReport)).asJava)
```
Comment (raboof) [inline src/main/scala/com/github/sbt/sbom/BomExtractor.scala:105]: does this make things much more heavy? if so we might want to put this behind an option - but I'm fine with just having it always-included for now.

File: src/main/scala/com/github/sbt/sbom/BomExtractor.scala [hunk 4]
```diff
@@ -95,22 +119,42 @@ class BomExtractor(settings: BomExtractorParams, report: UpdateReport, log: Logg
...
+
     private def licenseChoice: Option[LicenseChoice] = {
-      val licenses: Seq[model.License] = moduleReport.licenses.map { case (name, mayBeUrl) =>
-        model.License(name, mayBeUrl)
+      val licenses: Seq[License] = moduleReport.licenses.map { case (name, urlOption) =>
+        val license = new License()
+        urlOption.foreach { licenseUrl =>
+          /*TODO: LicensesArchive.bundled.findByNormalizedUrl(licenseUrl).foreach { archiveLicense =>
+            license.setId(archiveLicense.id)
+          }*/
```
Comment (raboof) [inline src/main/scala/com/github/sbt/sbom/BomExtractor.scala:142]: now I see why #94 :)

what are typical values for `name` here? would that be useful to distinguish between e.g. `GPL-2.0` and `GPL-2.0-or-later`?

**Review by raboof — COMMENTED** [review]
looks good, a few optional comments to consider, and you'll have to update the scripted test expectations to match the new behavior.

File: src/main/scala/com/github/sbt/sbom/BomSbtPlugin.scala [hunk 5]
```diff
@@ -25,6 +25,9 @@ object BomSbtPlugin extends AutoPlugin {
     lazy val includeBomSerialNumber: SettingKey[Boolean] = settingKey[Boolean](
       "should the resulting BOM contain a serial number? default is false, because the current mechanism for determining the serial number is not reproducible"
     )
+    lazy val includeBomHashes: SettingKey[Boolean] = settingKey[Boolean](
+      "should the resulting BOM contain artifact hashes? default is true"
```
Comment (raboof) [inline src/main/scala/com/github/sbt/sbom/BomSbtPlugin.scala:29]: I think you don't have that default specified yet

**Review by raboof — APPROVED** [review]
code is looking good, just needs the default value for 'include bom hashes' in the plugin and updated test expectations in the scripted tests

**Comment by lhns** [comment]
oh no why are the hashes different in ci...

**Comment by lhns** [comment]
ohh they are selected based on the java version in org.cyclonedx.util.BomUtils:
```java
        try {
            digests.add(DigestUtils.getSha3_256Digest());
        } catch (Exception | NoSuchMethodError e) { /* Not available in Java 8 and only available in later versions of DigestUtils */ }
```

File: src/main/scala/com/github/sbt/sbom/BomSbtPlugin.scala [hunk 6]
```diff
@@ -46,6 +58,10 @@ object BomSbtPlugin extends AutoPlugin {
       bomFileName := bomFileNameSetting.value,
       bomSchemaVersion := defaultSupportedVersion.getVersionString,
       includeBomSerialNumber := false,
+      includeBomTimestamp := false,
+      includeBomToolVersion := true,
+      includeBomHashes := true,
+      enableBomSha3Hashes := true,
```
Comment (raboof) [inline src/main/scala/com/github/sbt/sbom/BomSbtPlugin.scala:64]: I wonder if we should default this to `false`, to be more reproducible by default...

File: README.md [hunk 7]
```diff
@@ -45,9 +45,14 @@ The `listBom` command can be used to generate the contents of the BOM without wr
...
-|-------------|--------|---------------|
-| bomFileName | String | bom file name |
+| Setting                | Type    | Default                                      | Description                    |
+|------------------------|---------|----------------------------------------------|--------------------------------|
+| bomFileName            | String  | `"${artifactId}-${artifactVersion}.bom.xml"` | bom file name                  |
+| bomSchemaVersion       | String  | `"1.6"`                                      | bom schema version             |
+| includeBomSerialNumber | Boolean | `false`                                      | include serial number in bom   |
+| includeBomTimestamp    | Boolean | `false`                                      | include timestamp in bom       |
+| includeBomToolVersion  | Boolean | `true`                                       | include tool version in bom    |
+| includeBomHashes       | Boolean | `true`                                       | include artifact hashes in bom |
```
Comment (raboof) [inline README.md:55]: Maybe include `enableBomSha3Hashes` in this table?

**Review by raboof — COMMENTED** [review]
One final tweak, then looks great to me!

File: src/main/scala/com/github/sbt/sbom/BomSbtPlugin.scala (hunk 6 above)
Comment (lhns) [reply src/main/scala/com/github/sbt/sbom/BomSbtPlugin.scala:64]: I also thought about this but java 8 is mostly end of life anyway and the sha3 family will be pretty important in the future and should probably be enabled by default sooner than later. The gradle cyclonedx plugin also does not filter out hashes.

**Review by raboof — APPROVED** [review]
Great improvement!

**Pull request merged**